### ISSUE TITLE
Difficulty using gcc -shared to make a PowerPC shared library (shlb)

### ISSUE THREAD

**Issue by Raptor007** [autc04/Retro68#97]
Short version: `powerpc-apple-macos-gcc -shared` doesn't link my code into the xcoff.

I'm trying to compile aq2-tng for Mac Quake 2, which runs mod code from a PPC shared library.  The latest version of the mod uses a GNU Makefile to create gamei386.so / gamex86_64.so for Linux and gamex86.dll for Windows, so I'm working towards a gameppc.lib target for Mac.

Here's a simplified version of my build commands (in shell):

`RT="path/to/Retro68-build/toolchain"`
`$RT/bin/powerpc-apple-macos-gcc -O2 -c example.c -o example.o`
`$RT/bin/powerpc-apple-macos-gcc -O2 -shared example.o -o gameppc.xcoff`
`$RT/bin/MakePEF gameppc.xcoff -o gameppc.pef`
`$RT/bin/Rez -I$RT/powerpc-apple-macos/RIncludes $RT/powerpc-apple-macos/RIncludes/RetroPPCAPPL.r -DCFRAG_NAME="\"gameppc.lib\"" -o gameppc.bin --cc gameppc.lib -t shlb -c Q2id --data gameppc.pef`

It doesn't throw errors, but it generates a small xcoff file of about 200KB, a pef file of about 16KB, and a lib of about 20KB.  That's way too small, and I can see with a hex editor that none of those files contains any of the mod's function names.  They are present in the .o files.

If I remove `-shared` the linker wants `.main` defined.  A simple `int main(void) { return 0; }` resulted in a tiny xcoff again, likely because the linker optimized away all the unused objects, but having main call the mod's `GetGameAPI` function seems to do the trick.  Now gameppc.lib is about 680KB and I can see the expected code blocks in gameppc.lib with a hex editor or disassembler.

I tried putting `-shared` back onto the linker line with the dummy main, but it went back to tiny output files, perhaps discarding my main object like the others.  Should `-shared` be working?  Is there a better way to build a shlb file?

I'm not sure if I actually have a working shared library with this method.  Quake 2 gives a cryptic error message when I try to load the mod, which is the same thing I ran into after building the shared library in CodeWarrior 8.  Disassembling an old version of gameppc.lib that works, I see some additional code like `ImportMacFunctions` that is unfortunately documented nowhere that I've been able to discover.  Of course, that's a Mac Quake 2 issue, not a Retro68 issue.

**Comment by autc04** [comment]
Well, `-shared` should be the right way. It does look like the linker is removing unused code and not exporting anything.
It might be necessary to add some option to the linker command line...

**Comment by autc04** [comment]
Just found out / remembered that `MakePEF` does not support exporting symbols yet, so for now it just won't work (unless you use MPW's MakePEF tool inside an emulator instead).

To get your exported symbol into the XCOFF file, some options for `ld` will still be necessary (didn't get around to figuring them out exactly yet)

**Comment by autc04** [comment]
Try again now.
I have added (most of) the missing functionality to the `MakePEF` command.
To mark your functions as exported, put a list of exported functions in a text file (e.g., `gameppc.exp`), and use
```
powerpc-apple-macos-gcc -O2 -shared example.o -o gameppc.xcoff -Wl,-bE:gameppc.exp
```

**Comment by Raptor007** [comment]
Thank you for working on this!  I'm now as close to getting this working as I was with CodeWarrior 8, without having to revert the code to earlier C standards.  I'm still trying to reverse-engineer what Mac Quake2 needs in a mod, so I went with `-Wl,bexpfull` to make sure nothing gets left out.

I'm running Retro68 on Leopard with Tigerbrew (issue #96) so I had to tweak `target_link_options(...)` to `set_target_properties(... LINK_FLAGS ...)` in your sample.
